### PR TITLE
Reject modules as superclasses

### LINKED ISSUE #2252: Superclasses are not checked to be classes
#### Input

[→ View on sorbet.run](https://sorbet.run/#%23%20typed%3A%20true%0A%0Amodule%20A%3B%20end%0Aclass%20B%20%3C%20A%3B%20end%0A)

```ruby
# typed: true

module A; end
class B < A; end
```

#### Observed output

```
No errors! Great job.
```

#### Expected behavior

I would expect this to raise some kind of type error when `srb tc` is run, as subclassing a non-Class is always illegal (to my knowledge) and will throw an exception in the Ruby interpreter. I found this by forgetting to change `module` to `class` after refactoring an interface into an abstract class.

### PR DESCRIPTION
**Pull request by elliottt**
Reject classes declared with a superclass that is a module

### Motivation

Fixes #2252

### Test plan

See included automated tests.

### CODE REVIEW

**Comment by elliottt** [comment]
We have a policy of testing changes to Sorbet against Stripe's codebase before
merging them. I've kicked off a test run for the current PR.  When the build
finishes, I'll share with you whether or how it failed. Thanks!

Stripe employees can see the build results here:

→ https://go/builds/bui_Jqu8akXrojX7H7
→ https://go/builds/bui_Jqu91vHFEE1Gjd

File: definition_validator/validator.cc [hunk 1]
```diff
@@ -495,6 +495,39 @@ void validateSealed(core::Context ctx, const core::ClassOrModuleRef klass, const
     validateSealedAncestorHelper(ctx, singleton, classDef, klass, "extended");
 }
 
+void validateSuperClass(core::Context ctx, const core::ClassOrModuleRef sym, const ast::ClassDef &classDef) {
+    if (!sym.data(ctx)->isClassOrModuleClass()) {
+        return;
```
Comment (jez) [inline definition_validator/validator.cc:500]:
```suggestion
        ENFORCE(classDef.kind != ast::ClassDef::Kind::Class);
        return;
```

File: definition_validator/validator.cc [hunk 2]
```diff
@@ -495,6 +495,39 @@ void validateSealed(core::Context ctx, const core::ClassOrModuleRef klass, const
...
+    }
+
+    const auto superClass = sym.data(ctx)->superClass();
+    if (!superClass.exists()) {
+        return;
```
Comment (jez) [inline definition_validator/validator.cc:505]:
```suggestion
        // Happens for certain special classes at the top of the inheritance hierarchy.
        return;
```

File: test/testdata/lsp/symbol_query_filter__main.rb [hunk 3]
```diff
@@ -16,7 +16,7 @@ def isFoo
   end
 end
 
-class SomeClass
+class SomeClass # error: The super class `AnotherModule::NestedModule` of `SomeClass` does not derive from `Class`
```
Comment (aprocter-stripe) [inline test/testdata/lsp/symbol_query_filter__main.rb:19]: Extremely slight bummer that this error locates here when the actual offense is down below when we reopen the class. (But I guess one sees it down there as well, and I dunno if there's any way to make this not happen.)

File: test/testdata/resolver/module_superclass.rb [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+class SubclassModule2 < Module; end
+
+# It's okay to subclass Class
+class SubclassClass < Class; end
+
+module M; end
+
+# It's not okay to subclass a module
+  class SubclassModule1 < M; end
+# ^^^^^^^^^^^^^^^^^^^^^^^^^ error: The super class `M` of `SubclassModule1` does not
```
Comment (aprocter-stripe) [inline test/testdata/resolver/module_superclass.rb:19]:
```suggestion
# ^^^^^^^^^^^^^^^^^^^^^^^^^ error: The super class `M` of `SubclassModule1` does not derive from `Class`
```

**Review by aprocter-stripe — APPROVED** [review]
Good stuff!

File: test/testdata/lsp/symbol_query_filter__main.rb (hunk 3 above)
Comment (elliottt) [reply test/testdata/lsp/symbol_query_filter__main.rb]: Good point. I've added an additional check to exit early if the `classDef` is missing a superclass 👍

**Pull request merged**